### ISSUE TITLE
Make CI more resilient (& hermetic) to remote repo failures

### ISSUE THREAD

**Issue by aiuto** [bazelbuild/bazel#8911]
TL;DR;
In trying to test PR #8910 I am seeing failures on only a few out of many platforms. A test for building the distribution bootstrap fails because we get a spurious failure to download a distro.

Two questions/issues come to mind.
1. Why doe we run this kind of packaging tests on any system but the ones we build source boot distributions from?
  We could save CI time by not testing things on platforms we don't use for distribution
2. It seems like we want to pre-load new repos into a secure, high-performance repository so that we never have to go off machine during CI runs.
  We could automate the process of gathering requisites from WORKSPACE and pushing to the mirror.

More detail:

[1A[K
[1A[K[32mINFO: [0mCall stack for the definition of repository 'rules_pkg' which is a http_archive (rule definition at /tmp/bazel_MOkNwZwU/out/external/bazel_tools/tools/build_defs/repo/http.bzl:292:16):
 - /var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/ec321eb2cc2d0f8f91b676b6d4c66c29/sandbox/linux-sandbox/5715/execroot/io_bazel/_tmp/2a109a4137b457914dd3a6d3268626c4/bazelbootstrap.UNwURgbY/WORKSPACE:635:1
[32mLoading:[0m 0 packages loaded
    Fetching @rules_pkg; fetching

[1A[K
[1A[K[31m[1mERROR: [0mAn error occurred during the fetch of repository 'rules_pkg':
   java.io.IOException: Error downloading [https://github.com/bazelbuild/rules_pkg/releases/download/0.2.0/rules_pkg-0.2.0.tar.gz] to /tmp/bazel_MOkNwZwU/out/external/rules_pkg/rules_pkg-0.2.0.tar.gz: Unknown host: github.com
[32mLoading:[0m 0 packages loaded

Full log:
https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/4134ea2c-77b6-445d-9bbb-e3914d322cb9/src/test/shell/bazel/bazel_bootstrap_distfile_test/attempt_1.log

**Comment by aehlig** [comment]
> In trying to test PR #8910 I am seeing failures on only a few out of many platforms. A test for building the distribution bootstrap fails because we get a spurious failure to download a distro.

That download failure is not spurious; it is a [deliberate property of the test to trun off the network](https://github.com/bazelbuild/bazel/commit/f250784293167ee01dab778731573a36e7820499) to catch errors like the one introduced in PR #8910 (see my review there).

**Comment by aiuto** [comment]
Got it (now).  What confused me is that *some* platforms worked, but others did not.

I consolidated what you and Phillip explained yesterday here.
https://github.com/bazelbuild/bazel/wiki/Adding-a-new-dependency-on-an-external-repo

The second part of my issue is still worth exploring. Can we do a make the process of updating WORKSPACE from the archives more automated.  Straw man proposal

external_repo_list.py:

REPOS = [
    {"name": rules_foo-1.0.0.tgz",
     "sha256": "123031238r230rf223f2",
     "urls": [ url list ],
     "type": "http_archive"
   }
   ...
]

Now we create a helper tool which
- reads the list
- loads WORKSPACE.oss
- rewrites the distfiles rule
- rewrites the http_archive entry

The helper could also run in test mode to see that applying it would not modify WORKSPACE. Then we could make that part of the unit testing, and fail with a very clear explanation if we updating something wrong.
